File: tests/XRoadLib.Tests/XRoadFilterMapTest.fs
namespace XRoadLib.Tests

open FsUnit
open NUnit.Framework
open System
open XRoadLib.Serialization

[<TestFixture>]
module XRoadFilterMapTest =
    type CustomClass() =
        member val Member1 = "" with get, set
        member val Member2 = 0 with get, set
        member val Member3 = CustomClass() with get, set

    type NoCustomMap() = inherit XRoadFilterMap<CustomClass>("Group name")

    type CustomMap() as this =
        inherit XRoadFilterMap<CustomClass>("CustomMap")
        do
            this.Enable(fun x -> x.Member1)
            this.Enable(fun x -> x.Member3)
            this.Enable(fun x -> x.Member1)
            this.Enable(fun x -> x.Member1)

    type BrokenMap() as this =
        inherit XRoadFilterMap<CustomClass>("BrokenMap")
        do this.Enable(fun x -> x.Member3.Member1)

    type Broken2Map() as this =
        inherit XRoadFilterMap<CustomClass>("Broken2Map")
        do this.Enable(fun x -> true)

    [<Test>]
    let ``missing mappings`` () =
        let map = NoCustomMap() :> IXRoadFilterMap
        map.GroupName |> should equal "Group name"
        map.EnabledProperties |> should not' (be Null)
        map.EnabledProperties.Count |> should equal 0

    [<Test>]
    let ``ignore re-enabled members`` () =
        let map = CustomMap() :> IXRoadFilterMap
        map.GroupName |> should equal "CustomMap"
        map.EnabledProperties |> should not' (be Null)
        map.EnabledProperties.Count |> should equal 2
        map.EnabledProperties.Contains("Member1") |> should equal true
        map.EnabledProperties.Contains("Member3") |> should equal true

    [<Test>]
    let ``cannot handle nested properties`` () =
        TestDelegate(fun _ -> BrokenMap() |> ignore)
        |> should (throwWithMessage "Only parameter members should be used in mapping definition (BrokenMap).") typeof<ArgumentException>

    [<Test>]
    let ``cannot handle non-member expressions`` () =
        TestDelegate(fun _ -> Broken2Map() |> ignore)
        |> should (throwWithMessage "MemberExpression expected, but was ConstantExpression (Broken2Map).") typeof<ArgumentException>

File: test/XRoadLib.Tests/Serialization/XRoadFilterMapTest.cs
using System;
using XRoadLib.Schema;
using XRoadLib.Serialization;
using Xunit;

namespace XRoadLib.Tests.Serialization
{
    public class XRoadFilterMapTest
    {
        private class CustomClass
        {
            public string Member1 { get; set; } = "";
            public int Member2 { get; set; }
            public CustomClass Member3 { get; set; } = new CustomClass();
        }

        private class NoCustomMap : XRoadFilterMap<CustomClass>
        {
            public NoCustomMap()
                : base("Group name")
            { }
        }

        private class CustomMap : XRoadFilterMap<CustomClass>
        {
            public CustomMap()
                : base("CustomMap")
            {
                Enable(x => x.Member1);
                Enable(x => x.Member3);
                Enable(x => x.Member1);
                Enable(x => x.Member1);
            }
        }

        private class BrokenMap : XRoadFilterMap<CustomClass>
        {
            public BrokenMap()
                : base("BrokenMap")
            {
                Enable(x => x.Member3.Member1);
            }
        }

        private class Broken2Map : XRoadFilterMap<CustomClass>
        {
            public Broken2Map()
                : base("Broken2Map")
            {
                Enable(x => true);
            }
        }

        [Fact]
        public void MissingMappings()
        {
            var map = (IXRoadFilterMap)new NoCustomMap();
            Assert.Equal("Group name", map.GroupName);
            Assert.NotNull(map.EnabledProperties);
            Assert.Equal(0, map.EnabledProperties.Count);
        }

        [Fact]
        public void IgnoreReEnabledMembers()
        {
            var map = (IXRoadFilterMap)new CustomMap();
            Assert.Equal("CustomMap", map.GroupName);
            Assert.NotNull(map.EnabledProperties);
            Assert.Equal(2, map.EnabledProperties.Count);
            Assert.True(map.EnabledProperties.Contains("Member1"));
            Assert.True(map.EnabledProperties.Contains("Member3"));
        }

        [Fact]
        public void CannotHandleNestedProperties()
        {
            var exception = Assert.Throws<SchemaDefinitionException>(() => new BrokenMap());
            Assert.Equal("Only parameter members should be used in mapping definition (BrokenMap).", exception.Message);
        }

        [Fact]
        public void CannotHandleNonMemberExpressions()
        {
            var exception = Assert.Throws<SchemaDefinitionException>(() => new Broken2Map());
            Assert.Equal("MemberExpression expected, but was ConstantExpression (Broken2Map).", exception.Message);
        }
    }
}